### DISCUSSION TITLE
Run this program in CLI

### DISCUSSION THREAD

**Discussion by saadkhan955** [receyuki/stable-diffusion-prompt-reader#69] [Q&A]
Is it possible to run a program in the CLI to batch edit the metadata attached to PNG files? I've generated 576 images in Automatic1111 and want to upload them to CivitAI. However, when I try to upload, the auto-analyzing feature doesn't display the 'Lora' data for any of the images. I suspect this is due to incorrect hashes in the PNG metadata.

I've manually corrected the metadata for some files using 'SD_Prompt_reader', but this is tedious. Could the script be modified to run 'SD_Prompt_reader' in the CLI? Ideally, the script would read an updated metadata text file and then save the changes to the corresponding PNG files.

**Comment by receyuki** [comment]
I don't have plans to provide a command-line tool yet, but hey, this is python, and I've already published the SD Prompt Reader on PyPI. If you're not experienced with Python, I can provide some simple Python scripts to call the SD Prompt Reader for batch processing (you can even use chatgpt to write some simple scripts).

The new version that supports batch processing is already half done, but I haven't had enough time to complete it recently. But anyhow, you've made me realise I should at least provide some simpler ways for users to batch process or call SD Prompt Reader through CLI.

**Reply by saadkhan955** [reply]
I've used ChatGPT to extract data from all the files into a text file. I also used it to correct the hashes. However, when I try to write the data back to the images, I'm encountering errors. I would very much appreciate it if you can provide some sample scripts to help with this process.

**Reply by receyuki** [reply]
I changed my mind immediately after saying "I have no plans yet", and now I've started writing command line tools lol.

BTW, Do you have experience with the Python environment? For example, running some simple Python scripts on your computer?

**Reply by saadkhan955** [reply]
yes I have experience with running python like creating venv, installing requirement, etc

**Reply by receyuki** [reply]
Ensure your Python version is higher than 3.10.
Run `pip install sd-prompt-reader` to install sd-prompt-reader as a library in Python.

```Python
from sd_prompt_reader.image_data_reader import ImageDataReader

# The image file you want to read
image_path = "example/image.png"
# The save location for the new file
new_image = "example/image_edited.png"

# The prompt you want to write into the file
# Please note, the setting for A1111 must begin with "Steps:"
new_positive = "I'm positive"
new_negative = "I'm negative"
new_setting = "Steps: 0, Sampler: Magic, CFG scale: 0, Seed: 0"

data = f"""{new_positive}
Negative prompt: {new_negative}
{new_setting}"""

# This is how you read metadata
with open(image_path, "rb") as f:
    image_data = ImageDataReader(f)
    print(image_data.positive)
    print(image_data.negative)
    print(image_data.setting)

# Voila!
image_data.save_image(image_path, new_image, "PNG", data)
```

I will provide a better way afterwards.

**Comment by receyuki** [comment]
I just released a beta version with a CLI.
https://github.com/receyuki/stable-diffusion-prompt-reader#cli
You can install it using the following command.
`pip install --pre --upgrade sd-prompt-reader==1.3.5b1.dev1`
